contrib/hagcn/hagcn_layers.py

```diff
@@ -0,0 +1,214 @@
+"""K-Hop Layer and Adaptive Filter Module from https://arxiv.org/pdf/1706.09916.pdf"""
+
+from __future__ import division
+from __future__ import unicode_literals
+
+__author__ = "Vignesh Ram Somnath"
+__license__ = "MIT"
+
+import numpy as np
+import tensorflow as tf
+import os
+import sys
+import logging
+
+from deepchem.models.tensorgraph.layers import Layer, convert_to_layers
+from deepchem.models.tensorgraph import initializations, model_ops
+from deepchem.models.tensorgraph import activations
+
+
+class AdaptiveFilter(Layer):
+
+  def __init__(self,
+               num_nodes,
+               num_node_features,
+               batch_size=64,
+               k=1,
+               init='glorot_uniform',
+               combine_method='linear',
+               **kwargs):
+    """
+      Parameters
+      ----------
+      num_nodes: int
+        Number of nodes in the graph
+      num_node_features: int
+        Number of features per node in the graph
+      batch_size: int, optional
+        Batch size used for training
+      k: int, optional
+        k-hop neighborhood to look at
+      init: str, optional
+        Initialization method for the weights
+      combine_method: str, optional
+        How to combine adjacency matrix and node features
+
+    """
+
+    if combine_method not in ['linear', 'prod']:
+      raise ValueError('Combine method needs to be one of linear or product')
+    self.k = k
+    self.num_nodes = num_nodes
+    self.num_node_features = num_node_features
+    self.batch_size = batch_size
+    self.init = initializations.get(init)
+    self.combine_method = combine_method
+    super(AdaptiveFilter, self).__init__(**kwargs)
+
+  def _build(self):
+    if self.combine_method == "linear":
+      self.Q = self.init(
+          shape=(self.batch_size, self.num_nodes,
+                 self.num_nodes + self.num_node_features))
+    else:
+      self.Q = self.init(
+          shape=(self.batch_size, self.num_node_features, self.num_nodes))
+
+    self.trainable_weights = [self.Q]
+
+  @staticmethod
+  def pow_k(inputs, k=1):
+    """Computes the kth power of inputs, used for adjacency matrix"""
+    if k == 1:
+      return inputs
+    if k == 0:
+      return tf.ones(inputs.shape)
+
+    if k % 2 == 0:
+      first_half = AdaptiveFilter.pow_k(inputs, int(k / 2))
+      second_half = AdaptiveFilter.pow_k(inputs, int(k / 2))
```

Comment:
Just realised that the `AdaptiveFilter.pow_k` method is more efficient with a single recursion for the `k % 2 == 0` case.
Will add in the next commit with a model to run the same.